Mage/src/main/java/mage/abilities/mana/ManaOptions.java

```diff
@@ -406,61 +389,46 @@ public ManaOptions copy() {
-                            newCombinations = true; // repeat the while as long there are new combinations and usage is repeatable
-
-                            Mana moreValuable = Mana.getMoreValuableMana(currentManaCopy, newMana);
-                            if (newMana.equals(moreValuable)) {
-                                oldManaWasReplaced = true; // the new mana includes all possibilities of the old one, so no need to add it after return
-                                if (!currentMana.equalManaValue(currentManaCopy)) {
-                                    this.removeEqualMana(currentManaCopy);
-                                }
-                            }
-                            currentManaCopy = newMana.copy();
-                        }
+            currentManaCopy.setToMana(currentMana); // copy start mana because in iteration it will be updated
+            do { // loop for multiple usage if possible
+                newCombinations = false;
+
+                newMana.setToMana(currentManaCopy);
+                newMana.subtract(payCombination);
+                // Get the mana to iterate over.
+                // If manaToAdd is specified add it, otherwise add the mana produced by the mana ability
+                List<Mana> manasToAdd = (manaToAdd != null) ? Collections.singletonList(manaToAdd) : manaAbility.getNetMana(game, newMana);
+                for (Mana mana : manasToAdd) {
+                    newMana.add(mana);
+                    if (this.contains(newMana)) {
+                        continue;
                     }
-                } else {
-                    Mana newMana = currentManaCopy.copy();
-                    newMana.subtract(payCombination);
-                    newMana.add(manaToAdd);
-                    if (!isExistingManaCombination(newMana)) {
-                        this.add(newMana); // add the new combination
-                        newCombinations = true; // repeat the while as long there are new combinations and usage is repeatable                            
-                        Mana moreValuable = Mana.getMoreValuableMana(currentManaCopy, newMana);
-                        if (newMana.equals(moreValuable)) {
-                            oldManaWasReplaced = true; // the new mana includes all possible mana of the old one, so no need to add it after return
-                            if (!currentMana.equalManaValue(currentManaCopy)) {
-                                this.removeEqualMana(currentManaCopy);
-                            }
+
+                    this.add(newMana.copy()); // add the new combination
+                    newCombinations = true; // repeat the while as long there are new combinations and usage is repeatable
+
+                    if (newMana.isMoreValuableThan(currentManaCopy)) {
+                        oldManaWasReplaced = true; // the new mana includes all possible mana of the old one, so no need to add it after return
```

Comment:
You can replace this assignment with `return true;`. This will skip the remaining iterations of the loop without changing this function's return value. This will also let you get rid of the variable `oldManaWasReplaced` altogether.

Reply:
I will have to think about that one and see. My gut feeling is that I can't just return true since the code must generate all combinations until the criteria for the while loop is met. The change of state is equally important to the return value (possibly more so).

Reply:
`oldManaWasReplaced` is initialized with a value of `false` on line 393. In a nested `for` loop, it's conditionally set to `true` on line 423. Its value is ultimately returned on line 433 with no potential to be flipped back to `false` if it's been set to `true`.

Does this method have effects on the state of the game, or is it just used to do a `true`/`false` check? If the method's purpose is just for checking if a value's `true`/`false`, line 423 can be replaced with `return true;` and line 433 can be replaced with `return false;`.